packages/wavs/src/subsystems/trigger/streams/evm_stream/client/connection.rs

```diff
@@ -166,19 +324,106 @@ async fn connection_loop(
-                            tracing::error!("EVM connection lost from {endpoint}: {err:?}");
-                        } else {
-                            tracing::info!("EVM: disconnected {endpoint}");
+
+                        let mut forced_switch = false;
+
+                        loop {
+                            if let Some(priority_endpoint) = priority_endpoint.as_ref() {
+                                if force_switch_flag.load(Ordering::SeqCst)
+                                    && !*is_using_priority.read().await
+                                {
+                                    tracing::info!(
+                                        "EVM: force switching to priority endpoint at index {}",
+                                        priority_endpoint.index
+                                    );
+                                    forced_switch = true;
+                                    endpoint_idx = priority_endpoint.index;
+                                    current_backoff = Connection::BACKOFF_BASE;
+                                    failures_in_cycle = 0;
+                                    break;
+                                }
+                            }
+
+                            tokio::select! {
+                                _ = &mut shutdown_rx => {
+                                    tracing::info!("EVM: shutdown requested, exiting connection loop");
+                                    break 'connection_loop; // Exit the outer loop directly
+                                }
+                                _ = force_switch_notify.notified(), if priority_endpoint.is_some() => {
+                                    continue;
+                                }
+                                message = stream.next() => {
+                                    match message {
+                                        Some(Ok(Message::Text(msg))) => {
+                                            if let Err(e) = connection_data_tx
+                                                .send(ConnectionData::Text(msg.to_string()))
+                                            {
+                                                tracing::error!("Failed to send text message from {}: {}", endpoint, e);
+                                            }
+                                        }
+                                        Some(Ok(Message::Binary(msg))) => {
+                                            if let Err(e) = connection_data_tx
+                                                .send(ConnectionData::Binary(msg.to_vec()))
+                                            {
+                                                tracing::error!("Failed to send binary message from {}: {}", endpoint, e);
+                                            }
+                                        }
+                                        Some(Ok(Message::Ping(_))) => {
```

Comment:
we lost the note about how tungstenite handles this automatically. If it didn't, it would be a bug to do nothing here, please restore that comment